### ISSUE TITLE
java.lang.NoSuchMethodError when attempting to compare 2 different pdfs

### ISSUE THREAD

**Issue by iqbalmmz** [red6/pdfcompare#7]
I have 2 Pdfs which differ in page count and content. I compare them as such:
Boolean isEquals = new PdfComparator(pdf1, pdf2).compare().writeTo("./diff.pdf");.........line 82

but encounter this error:
Exception in thread "main" java.lang.NoSuchMethodError: org.apache.pdfbox.pdmodel.PDDocument.load(Ljava/io/InputStream;Lorg/apache/pdfbox/io/MemoryUsageSetting;)Lorg/apache/pdfbox/pdmodel/PDDocument;
	at de.redsix.pdfcompare.PdfComparator.compare(PdfComparator.java:155)
	at com.company.ctg.fps.automation.integration.main(FPS.java:82)

**Comment by iqbalmmz** [comment]
pdf1 and pdf1 are File objects

**Comment by finsterwalder** [comment]
That's strange. Could you please verify, which version of PdfBox you use in your project?
e.g. execute "mvn dependency:tree" and look for the pdfbox version.
You should use 2.0.6
That's the only explanation I can find, since the tests work finde and they execute that code path.

**Comment by iqbalmmz** [comment]
Thanks for your prompt reply. And apologies. According to my pom I am using <version>1.1.20</version>. That looks to be the problem. I will upgrade to version `2.0.6` and will report back the result

**Comment by iqbalmmz** [comment]
Hmm...I seem to be having an issue when I updated my pom dependency to have 
`
		<dependency>
			<groupId>de.redsix</groupId>
			<artifactId>pdfcompare</artifactId>
			<version>2.0.6</version>
		</dependency>
`

**Comment by iqbalmmz** [comment]
Sorry I misunderstood your comment. 

So after running mvn dependency:tree

Looks like I am running pdfbox version 2.0.6

```
[INFO] +- de.redsix:pdfcompare:jar:1.1.20:compile
[INFO] |  +- org.apache.pdfbox:pdfbox:jar:2.0.6:compile
[INFO] |  |  +- org.apache.pdfbox:fontbox:jar:2.0.6:compile
[INFO] |  |  \- commons-logging:commons-logging:jar:1.2:compile
[INFO] |  +- com.typesafe:config:jar:1.3.0:compile
[INFO] |  +- org.slf4j:slf4j-api:jar:1.7.22:compile
[INFO] |  \- ch.qos.logback:logback-classic:jar:1.2.1:compile
[INFO] |     \- ch.qos.logback:logback-core:jar:1.2.1:compile
```

There may be another issue

**Comment by PascalSchumacher** [comment]
>I'm using 2 pdf files which differ in page count and content. Seems this should have been covered.

As far as I know this is covered. I think scenario happened to us.

Can you post the complete output of `mvn dependency:tree`?

Looks like classpath contains another pdfbox version (maybe with the old "pdfbox" groupid)?

Or maybe you pdfbox jar is corrupted. Can you check it?

**Comment by finsterwalder** [comment]
Your use case is definitely covered. That's the whole point of PdfCompare. You can compare any two Pdfs. No matter how similar or different they are.

Your error says, that your loaded class PDDocument is missing a method, that is definitely there in version 2.0.6 of PdfBox.
So it is most likely your classpath contains another version of the PDDocument class.

**Comment by iqbalmmz** [comment]
I'm confident I have v2.0.6, here is my dependency tree:

```
[INFO] 
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ fpscompare ---
[INFO] com.intuit.ctg.fps.automation:fpscompare:jar:1.0.1-SNAPSHOT
[INFO] +- com.intuit.cg.fs.fps.qa:PECommon:jar:0.0.1-SNAPSHOT:compile
[INFO] +- com.intuit.ctg.taxengine.automation:tpsconv-java:jar:1.0.3-SNAPSHOT:compile
[INFO] |  +- org.apache.tika:tika-app:jar:1.7:compile
[INFO] |  |  +- org.apache.tika:tika-parsers:jar:1.7:compile
[INFO] |  |  |  +- org.apache.tika:tika-core:jar:1.7:compile
[INFO] |  |  |  +- org.gagravarr:vorbis-java-tika:jar:0.6:compile
[INFO] |  |  |  +- edu.ucar:netcdf:jar:4.2.20:compile
[INFO] |  |  |  |  +- edu.ucar:unidataCommon:jar:4.2.20:compile
[INFO] |  |  |  |  |  \- net.jcip:jcip-annotations:jar:1.0:compile
[INFO] |  |  |  |  \- commons-httpclient:commons-httpclient:jar:3.1:compile
[INFO] |  |  |  +- net.sourceforge.jmatio:jmatio:jar:1.0:compile
[INFO] |  |  |  +- org.apache.james:apache-mime4j-core:jar:0.7.2:compile
[INFO] |  |  |  +- org.apache.james:apache-mime4j-dom:jar:0.7.2:compile
[INFO] |  |  |  +- org.apache.commons:commons-compress:jar:1.8.1:compile
[INFO] |  |  |  +- org.tukaani:xz:jar:1.5:compile
[INFO] |  |  |  +- commons-codec:commons-codec:jar:1.9:compile
[INFO] |  |  |  +- org.bouncycastle:bcmail-jdk15:jar:1.45:compile
[INFO] |  |  |  +- org.bouncycastle:bcprov-jdk15:jar:1.45:compile
[INFO] |  |  |  +- org.apache.poi:poi:jar:3.11:compile
[INFO] |  |  |  +- org.apache.poi:poi-scratchpad:jar:3.11:compile
[INFO] |  |  |  +- org.apache.poi:poi-ooxml:jar:3.11:compile
[INFO] |  |  |  |  \- org.apache.poi:poi-ooxml-schemas:jar:3.11:compile
[INFO] |  |  |  |     \- org.apache.xmlbeans:xmlbeans:jar:2.6.0:compile
[INFO] |  |  |  +- org.ccil.cowan.tagsoup:tagsoup:jar:1.2.1:compile
[INFO] |  |  |  +- org.ow2.asm:asm-debug-all:jar:4.1:compile
[INFO] |  |  |  +- com.googlecode.mp4parser:isoparser:jar:1.0.2:compile
[INFO] |  |  |  |  \- org.aspectj:aspectjrt:jar:1.8.0:compile
[INFO] |  |  |  +- com.drewnoakes:metadata-extractor:jar:2.6.2:compile
[INFO] |  |  |  |  \- xerces:xercesImpl:jar:2.8.1:compile
[INFO] |  |  |  +- de.l3s.boilerpipe:boilerpipe:jar:1.1.0:compile
[INFO] |  |  |  +- rome:rome:jar:1.0:compile
[INFO] |  |  |  |  \- jdom:jdom:jar:1.0:compile
[INFO] |  |  |  +- org.gagravarr:vorbis-java-core:jar:0.6:compile
[INFO] |  |  |  +- com.googlecode.juniversalchardet:juniversalchardet:jar:1.0.3:compile
[INFO] |  |  |  +- com.uwyn:jhighlight:jar:1.0:compile
[INFO] |  |  |  \- com.pff:java-libpst:jar:0.8.1:compile
[INFO] |  |  +- org.apache.tika:tika-serialization:jar:1.7:compile
[INFO] |  |  |  \- com.google.code.gson:gson:jar:2.2.4:compile
[INFO] |  |  \- org.apache.tika:tika-xmp:jar:1.7:compile
[INFO] |  |     \- com.adobe.xmp:xmpcore:jar:5.1.2:compile
[INFO] |  +- com.intuit.ctg.taxengine.automation:xmlbuilder:jar:1.0.6:compile
[INFO] |  |  \- xalan:xalan:jar:2.7.2:compile
[INFO] |  |     \- xalan:serializer:jar:2.7.2:compile
[INFO] |  |        \- xml-apis:xml-apis:jar:1.3.04:compile
[INFO] |  +- commons-io:commons-io:jar:2.4:compile
[INFO] |  \- org.apache.maven.plugins:maven-resources-plugin:jar:2.7:compile
[INFO] |     +- org.apache.maven:maven-plugin-api:jar:2.2.1:compile
[INFO] |     +- org.apache.maven:maven-project:jar:2.2.1:compile
[INFO] |     |  +- org.apache.maven:maven-profile:jar:2.2.1:compile
[INFO] |     |  +- org.apache.maven:maven-artifact-manager:jar:2.2.1:compile
[INFO] |     |  |  \- backport-util-concurrent:backport-util-concurrent:jar:3.1:compile
[INFO] |     |  +- org.apache.maven:maven-plugin-registry:jar:2.2.1:compile
[INFO] |     |  \- org.codehaus.plexus:plexus-container-default:jar:1.0-alpha-9-stable-1:compile
[INFO] |     |     \- junit:junit:jar:3.8.1:compile
[INFO] |     +- org.apache.maven:maven-core:jar:2.2.1:compile
[INFO] |     |  +- org.apache.maven.wagon:wagon-file:jar:1.0-beta-6:runtime
[INFO] |     |  +- org.apache.maven:maven-plugin-parameter-documenter:jar:2.2.1:compile
[INFO] |     |  +- org.apache.maven.wagon:wagon-http-lightweight:jar:1.0-beta-6:compile
[INFO] |     |  |  \- org.apache.maven.wagon:wagon-http-shared:jar:1.0-beta-6:compile
[INFO] |     |  |     +- nekohtml:xercesMinimal:jar:1.9.6.2:compile
[INFO] |     |  |     \- nekohtml:nekohtml:jar:1.9.6.2:compile
[INFO] |     |  +- org.apache.maven.wagon:wagon-http:jar:1.0-beta-6:compile
[INFO] |     |  +- org.apache.maven.wagon:wagon-webdav-jackrabbit:jar:1.0-beta-6:runtime
[INFO] |     |  |  \- org.apache.jackrabbit:jackrabbit-webdav:jar:1.5.0:runtime
[INFO] |     |  |     \- org.apache.jackrabbit:jackrabbit-jcr-commons:jar:1.5.0:runtime
[INFO] |     |  +- org.apache.maven.reporting:maven-reporting-api:jar:2.2.1:compile
[INFO] |     |  |  +- org.apache.maven.doxia:doxia-sink-api:jar:1.1:compile
[INFO] |     |  |  \- org.apache.maven.doxia:doxia-logging-api:jar:1.1:compile
[INFO] |     |  +- org.apache.maven.wagon:wagon-provider-api:jar:1.0-beta-6:compile
[INFO] |     |  +- org.apache.maven:maven-repository-metadata:jar:2.2.1:compile
[INFO] |     |  +- org.apache.maven:maven-error-diagnostics:jar:2.2.1:compile
[INFO] |     |  +- commons-cli:commons-cli:jar:1.2:compile
[INFO] |     |  +- org.apache.maven.wagon:wagon-ssh-external:jar:1.0-beta-6:runtime
[INFO] |     |  |  \- org.apache.maven.wagon:wagon-ssh-common:jar:1.0-beta-6:compile
[INFO] |     |  +- org.apache.maven:maven-plugin-descriptor:jar:2.2.1:compile
[INFO] |     |  +- org.codehaus.plexus:plexus-interactivity-api:jar:1.0-alpha-4:compile
[INFO] |     |  +- org.apache.maven.wagon:wagon-ssh:jar:1.0-beta-6:compile
[INFO] |     |  |  \- com.jcraft:jsch:jar:0.1.38:compile
[INFO] |     |  +- classworlds:classworlds:jar:1.1:compile
[INFO] |     |  \- org.sonatype.plexus:plexus-sec-dispatcher:jar:1.3:compile
[INFO] |     |     \- org.sonatype.plexus:plexus-cipher:jar:1.4:compile
[INFO] |     +- org.apache.maven:maven-artifact:jar:2.2.1:compile
[INFO] |     +- org.apache.maven:maven-settings:jar:2.2.1:compile
[INFO] |     +- org.apache.maven:maven-model:jar:2.2.1:compile
[INFO] |     +- org.apache.maven:maven-monitor:jar:2.2.1:compile
[INFO] |     +- org.codehaus.plexus:plexus-utils:jar:3.0.15:compile
[INFO] |     +- org.apache.maven.shared:maven-filtering:jar:1.2:compile
[INFO] |     |  +- org.apache.maven.shared:maven-shared-utils:jar:0.3:compile
[INFO] |     |  |  \- com.google.code.findbugs:jsr305:jar:2.0.1:compile
[INFO] |     |  \- org.sonatype.plexus:plexus-build-api:jar:0.0.4:compile
[INFO] |     \- org.codehaus.plexus:plexus-interpolation:jar:1.19:compile
[INFO] +- com.testautomationguru.pdfutil:pdf-util:jar:0.0.2:compile
[INFO] |  +- org.apache.pdfbox:pdfbox:jar:2.0.6:compile
[INFO] |  |  +- org.apache.pdfbox:fontbox:jar:2.0.6:compile
[INFO] |  |  \- commons-logging:commons-logging:jar:1.2:compile
[INFO] |  \- org.apache.pdfbox:pdfbox-tools:jar:2.0.6:compile
[INFO] |     \- org.apache.pdfbox:pdfbox-debugger:jar:2.0.6:compile
[INFO] +- de.redsix:pdfcompare:jar:1.1.20:compile
[INFO] |  +- com.typesafe:config:jar:1.3.0:compile
[INFO] |  +- org.slf4j:slf4j-api:jar:1.7.22:compile
[INFO] |  \- ch.qos.logback:logback-classic:jar:1.2.1:compile
[INFO] |     \- ch.qos.logback:logback-core:jar:1.2.1:compile
[INFO] +- org.testng:testng:jar:6.11:compile
[INFO] |  +- com.beust:jcommander:jar:1.64:compile
[INFO] |  \- org.yaml:snakeyaml:jar:1.17:compile
[INFO] \- org.slf4j:slf4j-log4j12:jar:1.6.2:compile
[INFO]    \- log4j:log4j:jar:1.2.16:compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.709 s
```

**Comment by finsterwalder** [comment]
The dependency tree looks ok. But how do you execute the test? Are you sure the correct jars are on the classpath there too?
I have really no other ideas.

**Comment by iqbalmmz** [comment]
I'm executing the test from a main method. I'm certain the correct jars are loaded in the class path but I have not tested this. Do you have any suggestions as how I can test if the correct jars are loaded on my classpath?

**Comment by finsterwalder** [comment]
That very much depends on the way you execute your test.
Do you run it in an IDE?
Do you run it on the commandline? If so, how?
Or what else?

**Comment by PascalSchumacher** [comment]
Looks like Apache Tika versions below 1.13 (version 1.7 is in the dependency tree above) included PDFBox 1.x and can not be used with PDFBox 2.x (I guess PDFBox is repakaged in Tika jar). See https://stackoverflow.com/questions/37383999/apache-tika-and-apache-pdfbox-2-0 and https://issues.apache.org/jira/browse/TIKA-1285 for details.

**Comment by iqbalmmz** [comment]
Thanks. I'm running this on Eclipse IDE (v Neon) on MacOS. I use Run configuration to pass in arguments to the main method. All my dependencies are loaded via maven. 

When I encountered this issue with your tool, I actually began looking at using apache pdfbox library directly and it worked great. I preferred to use your tool since it's much simpler

**Comment by iqbalmmz** [comment]
@PascalSchumacher thanks I will look into this.

**Comment by iqbalmmz** [comment]
@PascalSchumacher @finsterwalder it worked!!! Thank you both for the support you provided. 

Just out of curiosity what made you look into tika? Was it because upon inspection of the dependency tree you noticed it used pdfBox?

**Comment by PascalSchumacher** [comment]
The most probable reason for the NoSuchMethodError thrown by a PDFBox class is incompatible PDFBox versions in the class-path. The dependency tree output displayed only the 2.x version used by pdfcompare, but 1.x could be hiding somewhere. When I took a look at the other dependencies of your project I noticed Tika, which can work with PDF files. Therefore I googled "Tika PDFBox" and found the stackoverflow entry.
